support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -233,3 +250,81 @@ object CreateZuoraSubscription {
     }
 
 }
+
+object DigitalSubscriptionGiftRedemption {
+
+  def ifDigitalSubscriptionGiftRedemption(product: ProductType, paymentMethod: Either[PaymentMethod, RedemptionData]) = {
+    product match {
+      case d: DigitalPack if paymentMethod.isRight && d.readerType == Gift => paymentMethod.right.toOption
+      case _ => None
+    }
+  }
+
+  def redeemGift(
+    redemptionData: RedemptionData,
+    gifteeUserId: String,
+    requestInfo: RequestInfo,
+    state: CreateZuoraSubscriptionState,
+    zuoraService: ZuoraService,
+    catalogService: CatalogService
+  ): Future[HandlerResult[SendThankYouEmailState]] = {
+    for {
+      giftSubscriptionFields <- zuoraService.getSubscriptionFromRedemptionCode(redemptionData.redemptionCode)
+      validatedSubscription <- Future.fromTry(validateRedemptionData(giftSubscriptionFields))
+      fullGiftSubscription <- zuoraService.getSubscriptionById(validatedSubscription.id)
```

Comment:
is this second call only needed to get the rate plan so that we can find out the billing period?

Reply:
Yes

Reply:
thought so, fair enough.
Is there anything else we need to do to ensure idempodency?  should we be updating the CreatedInRequestId custom field when we redeem, and then checking it before rejecting as already redeemed?

Reply:
I'm thinking of the case when zuora is V slow but succeeding, so the step function times out and retries, but actually is succeding slowly.

Reply:
That is a good point, I'll investigate 👍